### PR TITLE
Fail hex decoding if string contains non-hex characters

### PR DESCRIPTION
**Pull request by tristan**
Previously it would still return something if non-hex characters were used.
If there is a better way of doing regex, let me know, this looks horrible! 🙈

### CODE REVIEW

File: EtherealCereal/EtherealCereal.swift [hunk 1]
```diff
@@ -118,14 +118,23 @@ public extension Data {
...
-            str = self
+
+        let regex: NSRegularExpression
+        do {
+            regex = try NSRegularExpression(pattern: "^(?:0x)?([a-fA-F0-9]*)$", options: .caseInsensitive)
+        } catch {
+            fatalError("Invalid regular expression pattern")
         }
+        let matches = regex.matches(in: self, range: NSMakeRange(0, self.count))
+        if (matches.count != 1) {
```
Comment (designatednerd) [inline EtherealCereal/EtherealCereal.swift]: in terms of Swift's idioms, this is probably better written as: 
```swift
guard matches.count == 1 else { return nil }
```

Or you could even combine the two guard statements (forgive me if this formatting is incorrect, I haven't had a chance to read the code standards yet) 

```swift
guard 
    matches.count == 1, 
    let firstMatch = matches.first else {
        return nil 
}
```

File: EtherealCereal/EtherealCereal.swift [hunk 2]
```diff
@@ -118,14 +118,23 @@ public extension Data {
...
-        if self.substring(to: offset) == "0x" {
-            str = self.substring(from: offset)
-        } else {
-            str = self
+
+        let regex: NSRegularExpression
+        do {
+            regex = try NSRegularExpression(pattern: "^(?:0x)?([a-fA-F0-9]*)$", options: .caseInsensitive)
+        } catch {
+            fatalError("Invalid regular expression pattern")
```
Comment (designatednerd) [inline EtherealCereal/EtherealCereal.swift]: Do you want this to stop the program entirely if the regex doesn't compile, even in production? If not, consider using `assertionFailure`, which no-ops in production. If you'd rather it fail SUPER loudly, then this works.

File: Tests/Tests.swift [hunk 3]
```diff
@@ -59,4 +59,11 @@ class Tests: XCTestCase {
         let expected: [UInt8] = [16, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 242]
         XCTAssertEqual(hexString.hexadecimalData!, Data(expected))
     }
+
+    func testBadHexData() {
+        let hexString = "abcdog"
+        if let _ = hexString.hexadecimalData {
```
Comment (designatednerd) [inline Tests/Tests.swift]: You can probably just `XCTAssertNil` here instead of the if statement.

File: EtherealCereal/EtherealCereal.swift (hunk 2 above)
Comment (tristan) [reply EtherealCereal/EtherealCereal.swift]: the pattern is static, and if it doesn't compile then a lot of core functionality will not work, so i think failing loudly is a good idea.

But i'd also prefer it not need to be compiled with every call. in other languages i'd compile the regex at boot time, and reuse it each time. would this make sense in swift as well?

e.g. something like this declared outside of the class:
`let HEX_DATA_RE = try ! NSRegularExpression(pattern: "^(?:0x)?([a-fA-F0-9]*)$", options: .caseInsensitive)`
(not sure about the syntax here, or the `try!`)

File: EtherealCereal/EtherealCereal.swift [hunk 4]
```diff
@@ -117,15 +117,26 @@ public extension Data {
     }
 }
 
+struct AppRegex {
+    static var hexadecimalDataRegex: NSRegularExpression = {
+        return try! NSRegularExpression(pattern: "^(?:0x)?([a-fA-F0-9]*)$", options: .caseInsensitive)
+    }()
```
Comment (designatednerd) [inline EtherealCereal/EtherealCereal.swift]: 👍

**Comment by designatednerd** [comment]
LGTM but someone with eyes on the rest of the codebase should probably 👍 this as well

**Comment by Elland** [comment]
Would love some improvements to naming, otherwise looks good to me to. 🤔

**Comment by tristan** [comment]
changing some naming is fine by me @Elland but you'll have to point out what it bad and offer suggestions as I'm obviously blind to it

**Pull request merged**